demos/filebench/workloads/oltp.f

```diff
@@ -0,0 +1,86 @@
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or http://www.opensolaris.org/os/licensing.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+#
+# Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
+# Use is subject to license terms.
+#
+
+# Workload "oltp" is not fully-support on Occlum.
+# Semaphore-related flowop should be commented before run.
+set $dir=/tmp
+set $eventrate=0
+set $iosize=2k
+set $nshadows=2
+set $ndbwriters=100
+set $usermode=200000
+set $filesize=800m
+set $memperthread=1m
+set $workingset=0
+set $logfilesize=10m
+set $nfiles=10
+set $nlogfiles=1
+set $directio=0
+eventgen rate = $eventrate
+
+# Define a datafile and logfile
+define fileset name=datafiles,path=$dir,size=$filesize,entries=$nfiles,dirwidth=1024,prealloc=100,reuse
+define fileset name=logfile,path=$dir,size=$logfilesize,entries=$nlogfiles,dirwidth=1024,prealloc=100,reuse
+
+define process name=lgwr,instances=1
+{
+  thread name=lgwr,memsize=$memperthread
+  {
+    flowop aiowrite name=lg-write,filesetname=logfile,
+        iosize=256k,random,directio=$directio,dsync
+    flowop aiowait name=lg-aiowait
+    flowop semblock name=lg-block,value=3200,highwater=1000
+  }
+}
+
+# Define database writer processes
+define process name=dbwr,instances=$ndbwriters
+{
+  thread name=dbwr,memsize=$memperthread
+  {
+    flowop aiowrite name=dbwrite-a,filesetname=datafiles,
+        iosize=$iosize,workingset=$workingset,random,iters=100,opennext,directio=$directio,dsync
+    flowop hog name=dbwr-hog,value=10000
+    flowop semblock name=dbwr-block,value=1000,highwater=2000
+    flowop aiowait name=dbwr-aiowait
+  }
+}
+
+define process name=shadow,instances=$nshadows
+{
+  thread name=shadow,memsize=$memperthread
+  {
+    flowop read name=shadowread,filesetname=datafiles,
+        iosize=$iosize,workingset=$workingset,random,opennext,directio=$directio
+    flowop hog name=shadowhog,value=$usermode
+    flowop sempost name=shadow-post-lg,value=1,target=lg-block,blocking
+    flowop sempost name=shadow-post-dbwr,value=1,target=dbwr-block,blocking
+    flowop eventlimit name=random-rate
+  }
+}
```

Comment:
I thought multi-process workloads cannot be supported since Occlum does not have vfork. So how can Occlum run such multi-process workloads?

Reply:
I found `vfork` implementation both in Occlum and NGO. And tests go well under multi-process configs after I replace `fork` to `vfork`.
And `vfork` is a "must-option" since even if we set `instances=1`, filebench main process still forks one child to run the test.

Reply:
Replacing to `vfork` can provide concurrency since filebench use `vfork`+`execve`.
[vfork man page](https://man7.org/linux/man-pages/man2/vfork.2.html):

> the calling thread is suspended until the child terminates (either normally, by calling _exit(2), or abnormally, after delivery of a fatal signal), or it makes a call to **execve(2)**.

Reply:
I see. I thought filebench uses fork alone, without execve. Are you sure the replacing fork with vfork is correct? I mean vfork has some limitations (a child process cannot modify the parent's variables). If so, then filebench should have a complete multi-process support running on Occlum, right? And if yes, then we don't need to rewrite .f files to avoid multi-process usage.

Reply:
When I enable DEBUG mode, log shows multi-process executes operations concurrently and the results are normally consistent with results on Linux.
And yes, no need to rewrite this part in .f files.